dockstore-client/src/main/java/io/dockstore/client/cli/YamlVerifyUtility.java

```diff
@@ -0,0 +1,182 @@
+        List<String> missingFiles = new ArrayList<>();
+
+        // Check Workflows
+        List<YamlWorkflow> workflows = dockstoreYaml12.getWorkflows();
+        if (!workflows.isEmpty()) {
+            for (YamlWorkflow workflow : workflows) {
+                List<String> filePaths = workflow.getTestParameterFiles();
+                filePaths.add(workflow.getPrimaryDescriptorPath());
+                missingFiles.addAll(filesExist(filePaths, basePath));
+            }
+        }
+        // Check Tools
+        List<YamlWorkflow> tools = dockstoreYaml12.getTools();
+        if (!tools.isEmpty()) {
+            for (YamlWorkflow tool : tools) {
+                List<String> filePaths = tool.getTestParameterFiles();
+                filePaths.add(tool.getPrimaryDescriptorPath());
+                missingFiles.addAll(filesExist(filePaths, basePath));
+            }
+        }
+
+        // Check service
+        Service12 service = dockstoreYaml12.getService();
+        if (service != null) {
+            missingFiles.addAll(filesExist(service.getFiles(), basePath));
+        }
+        if (!missingFiles.isEmpty()) {
+            StringBuilder errorMsgBuild = new StringBuilder();
+            for (String missingFile: missingFiles) {
+                errorMsgBuild.append(missingFile + '\n');
+            }
+            String errorMsg = errorMsgBuild.toString();
+            throw new ValidateYamlException(INVALID_FILE_STRUCTURE + errorMsg);
+        }
+    }
+
+    public static void dockstoreValidate(String path) throws ValidateYamlException {
+        Path workflowPath = Paths.get(path);
+
+        // Verify that the path is a valid directory
+        if (!Files.isDirectory(workflowPath)) {
+            throw new ValidateYamlException(INVALID_DOCKSTORE_YML + workflowPath.toString() + FILE_DOES_NOT_EXIST);
+        }
+
+        // Verify that .dockstore.yml exists
+        Path dockstoreYmlPath = Paths.get(path, DOCKSTOREYML);
+        if (!Files.exists(dockstoreYmlPath)) {
+            throw new ValidateYamlException(INVALID_DOCKSTORE_YML + dockstoreYmlPath.toString() + FILE_DOES_NOT_EXIST);
+        }
+
+        // Verify that .dockstore.yml is non-empty
+        File dockstoreYml = new File(dockstoreYmlPath.toString());
+        if (dockstoreYml.length() == 0) {
+            throw new ValidateYamlException(INVALID_DOCKSTORE_YML + dockstoreYmlPath.toString() + EMPTY_FILE);
+        }
+
+        // Load dockstoreYml into string
+        String dockstoreYmlString = "";
+        try {
+            dockstoreYmlString = Files.readString(dockstoreYmlPath);
+        } catch (IOException ex) {
+            // Unlikely to ever catch
+            throw new ValidateYamlException("Error reading " + dockstoreYmlPath.toString() + " :\n" + ex.getMessage());
+        }
+
+        // Verify that the Yaml is valid, however does not verify it is valid for dockstore
+        Yaml yaml = new Yaml();
```

Comment:
It's a bit of stretch for an attack, since it's not running on any of our servers, but I guess in theory somebody could create a malicious dockstore.yml that you fork/pull locally, so I think you should use SafeConstructor like here: https://github.com/dockstore/dockstore/blob/419dd80a42caf46691920c18cee516e371e6c0ec/dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYamlHelper.java#L208

@david4096 Thoughts on the need to do this in the client?

Reply:
Corrected, the line is now
```
       final Yaml safeYaml = new Yaml(new SafeConstructor());
        try {
            safeYaml.load(Files.readString(dockstoreYmlPath));
            out(dockstoreYmlPath + VALID_YAML_ONLY);
        } catch (Exception ex) {
            throw new ValidateYamlException(INVALID_DOCKSTORE_YML + dockstoreYmlPath.toString() + INVALID_YAML + ex.getMessage());
        }
```